### ISSUE TITLE
Mutable references can be used after immutable ones are assigned

### ISSUE THREAD

**Issue by rlavande** [rust-lang/rust#59728]
Hello, I'm new to the community. I was learning Rust from the Book and came across an odd problem.

This example is as given in the Rust book 

```
fn main() {
    let mut s = String::from("hello world");

    let word = first_word(&s);

    s.clear();
}

fn first_word(s: &String) -> &str {
    let bytes = s.as_bytes();

    for (i, &item) in bytes.iter().enumerate() {
        if item == b' ' {
            return &s[0..i];
        }
    }

    &s[..]
}
```

The compiler should give an error because 'word' lives on even after the last statement in main where then it goes out of scope but 'clear()' requires a mutable reference to 's' . So this should give a 'immutable after mutable' kind of error.

However this piece of code is not giving any errors.

I thought it might be because of the way the local variables are popped from the stack (that they would be in order of definition rather than the usual reverse order) but that should make no difference; that will happen only after the last statement. The last statement in 'main' however should be where the error lies. 

I don't know if it is because of compiler optimization of code or if it is a real issue.
Please respond.

**Comment by RReverser** [comment]
This is something that used to be an error before, but got fixed / improved with non-lexical lifetimes in Rust 2018.

Now compiler is smart enough to see that `word` is actually not used anymore by the time you use `.clear()`, and so such usage sequence is valid.

**Comment by jonas-schievink** [comment]
Yes, this is non-lexical lifetimes working as intended. Closing.

**Comment by rlavande** [comment]
Thanks for the quick reply!!
